vizro-core/src/vizro/models/_utils.py

```diff
@@ -0,0 +1,122 @@
+    "from vizro.managers import data_manager",
+    "from vizro.models.types import capture", #TODO: could make conditional based on content
+}
+
+
+def _get_import_statements(captured_info: List[CapturedCallableInfo]) -> Set[str]:
+    import_paths = set()
+    for info in captured_info:
+        for replacement in REPLACEMENT_STRINGS:
+            if replacement.detect_path in info.module:
+                import_paths.add(replacement.from_import(replacement.detect_path, info.name))
+    return import_paths
+
+
+def _get_callable_code_strings(captured_info: List[CapturedCallableInfo]) -> Set[str]:
+    code_strings = set()
+    for info in captured_info:
+        if info.code is not None:
+            code_strings.add(info.code)
+    return code_strings
+
+
+def _get_data_manager_code_strings(captured_info: List[CapturedCallableInfo]) -> Set[str]:
+    return {
+        f'# data_manager["{arg[1]}"] = ===> Fill in here <==='
+        for info in captured_info
+        for arg in info.args
+        if arg[0] == "data_frame"
+    }
+
+
+def _clean_module_string(module_string: str) -> str:
+    return next(
+        (replacement.replace_path for replacement in REPLACEMENT_STRINGS if replacement.detect_path in module_string),
+        "",
+    )
+
+
+def _repr_clean(info: CapturedCallableInfo) -> str:
+    """Alternative __repr__ method with cleaned module paths."""
+    args = ", ".join(f"{key}={value!r}" for key, value in info.args)
+    module_path = f"{info.module}"
+    modified_module_path = _clean_module_string(module_path)
+    x = f"{modified_module_path}{info.name}({args})"
+    return x
+
+
+def _dict_to_python(model_data: Any, captured_info: Optional[List[CapturedCallableInfo]] = None) -> str:
+    """Function to generate python string from pydantic model dict."""
+    from vizro.models.types import CapturedCallable  # TODO: can we get rid of this?
+
+    if captured_info is None:
+        captured_info = []
+
+    if isinstance(model_data, Dict):
+        if MODEL_NAME in model_data:
+            model_name = model_data.pop(MODEL_NAME)
+            if model_name == ACTIONS_CHAIN:
+                action_data = model_data[ACTION]
+                if isinstance(action_data, List):
+                    return ", ".join(_dict_to_python(item, captured_info) for item in action_data)
+                else:
+                    return _dict_to_python(action_data, captured_info)
+            else:
+                other_content = ", ".join(
+                    f"{key}={_dict_to_python(value, captured_info)}" for key, value in model_data.items()
+                )
+                return f"vm.{model_name}({other_content})"
+        else:
+            return ", ".join(f"{key}={_dict_to_python(value, captured_info)}" for key, value in model_data.items())
+    elif isinstance(model_data, List):
+        return "[" + ", ".join(_dict_to_python(item, captured_info) for item in model_data) + "]"
+    elif isinstance(model_data, CapturedCallable):
```

Comment:
Hm very interesting - I am sure we can get around this because we do have the code for this. I suspect the root cause of this is because it is never imported so that the inspect module cannot find the source.

Reply:
This is ressolved now